### ISSUE TITLE
[Feature Request] New `BeartypeConf.claw_decorator_position` option for justice, victory, and user sanity

### ISSUE THREAD

**Issue by leycec** [beartype/beartype#374]
This feature request comes courtesy bestest @beartype users @danielward27 and @patrick-kidger at [parent feature request #368](https://github.com/beartype/beartype/issues/368#issuecomment-2074276090).

Previously, everybody hated that the `@beartype` decorator *had* to be applied first (i.e., bottommost) in decorator chains involving `@beartype` and one or more competing decorators that `@beartype` probably can't stand but is learning to tolerate with grace. <sup>*Decorators. They're like people. Somehow.*</sup>

Now, everybody hates that `beartype.claw.beartype_*()` import hooks forcefully inject the `@beartype` decorator last (i.e., topmost) in decorator chains. Everybody isn't wrong. Everybody is right. You know best. @leycec knows worst, because @leycec doesn't know what *you* know. If @leycec did, it'd be even harder for @leycec to sleep at night. @leycec is now speaking about himself in the third person. You know it's *really* late at night when...

## WAIT. What Were We Talking About Again?

Let's define a new `BeartypeConf.claw_decorator_position` option enabling you to configure the position of the `@beartype` decorator injected by `beartype.claw.beartype_*()` import hooks into decorator chains. Some users might prefer @beartype confine itself to the "pure" world of pure-Python functions by injecting `@beartype` first; others might prefer the current "impure" world of Anything Goes Beartype by injecting `@beartype` last.

Let's define this as an extensible enumeration rather than a simple boolean. Who knows what the future might hold? It probably holds pain, but let's be optimistic here. This enumeration might resemble:

```python
from enum import (
    Enum,
    auto as next_enum_member_value,
    unique as die_unless_enum_member_values_unique,
)

@die_unless_enum_member_values_unique
class BeartypeDecoratorPosition(Enum):
    '''
    Enumeration of all kinds of :func:`beartype.beartype` **decorator
    positions** (i.e., competing locations for :mod:`beartype.claw`
    import hooks to inject the :func:`beartype.beartype` decorator into
    existing decorator chains of one or more decorators decorating classes
    and callables to be additionally decorated by :func:`beartype.beartype`,
    each with concomitant tradeoffs with respect to decorator
    interoperability and quality assurance).

    Attributes
    ----------
    FIRST : EnumMemberType
        **First decorator position** (i.e, injecting
        :func:`beartype.beartype` as the first decorator in all decorator
        chains).
    LAST : EnumMemberType
        **Last decorator position** (i.e., injecting
        :func:`beartype.beartype` as the last decorator in all decorator
        chains). As the default, this position need *not* be explicitly
        enabled.
    '''

    FIRST = next_enum_member_value()
    LAST = next_enum_member_value()
```

Let's then define a new `BeartypeConf.claw_decorator_position` option, accepted as a new optional `BeartypeConf(..., claw_decorator_position: BeartypeDecoratorPosition = BeartypeDecoratorPosition.LAST)` parameter.

Let's. By "Let's," I *actually* mean: "Please. Gods. Do this for me! Somebody! Help! I'm begging you here!!!"

## WAIT. Why Is `BeartypeDecoratorPosition.LAST` the Default? I Hate That. Feel Bad, You.

**Look.** I already feel bad. I'm bald. Once, I had hair. It doesn't get worse. Things can only get better from here. Seriously, though. @leycec now [sinks lower by quoting himself](https://github.com/beartype/beartype/issues/368#issuecomment-2074276090):

@beartype is a gluttonous pig. The more @beartype knows, the better @beartype behaves. This is why `BeartypeDecoratorPosition.LAST` is the default.

@beartype wants full-spectrum runtime knowledge about everything that's *actually* happening – not the idealistic world of pure-Python functions that were originally defined by the user, but the *actual* on-the-ground nitty-gritty possibly-C-or-Rust-or-Fortran-or-*whatevah* objects and types that are *actually* being run. If an intermediary decorator wants to mutate, destroy, or add new parameters or parameter type hints, @beartype wants to respect that. If an intermediary decorator wants to fundamentally change the basic nature of physical reality, @beartype wants to know about that. @beartype wants it all. @beartype is like the U.S. government that way. I say this because I am now paying expat taxes. I laugh even as I cry. Mostly, I cry. :sob:

It's not just knowledge, though – it's also safety. @beartype *probably* shouldn't forcefully inject itself into workflows that it wasn't intended to be injected into. Should *all* type-checking wrapper functions dynamically generated by @beartype be automatically JITted by JIT decorators like `@numba.jit`, `@jax.jit`, or `@equinox.filter_jit`? Because that's what would happen under `beartype_this_package()` import hooks if `@beartype` were applied at the bottom rather than the top.

In the case of at least `@numba.jit`, the answer is almost certainly: "@beartype should *never* do that." `@numba.jit` is incredibly fragile and can't necessarily cope with all of the insane runtime logic that @beartype generates (e.g., extremely nested assignment statements that just go on and on for hundreds of lines).

It's not just knowledge or safety, though – it's also generality. Users should be able to apply the `@beartype` decorator wherever they like, to whatever they like, in whatever order they like. Originally, `@beartype` was *super*-strict and only wanted to be the bottom-most decorator in any decorator chain. But everybody hated that and complained about that, because *nobody* <sup>including myself, honestly</sup> could remember where in any given decorator chain `@beartype` was supposed to go. Now `@beartype` can go anywhere. Nobody has to think about it. The world is already too complex and burdensome. @beartype wants to reduce that cognitive load. This is how we reduce that cognitive load.

That said, I see both sides. Both sides have merit here... *Oh, who'm I kidding!?* One side rocks. But you hate that side. Grandpa Simpson, help us!

![](https://media1.tenor.com/m/NFNLxDTKlKIAAAAC/grandpa-both.gif)

**Comment by danielward27** [comment]
Another quick thought on this issue. Importing at the top I imagine only partially solves the issue for ``@decorator_that_needs_to_be_bottom``. If you use a package wide import hook, if you call ``my_inner_func`` (typechecked from the hook) from within ``my_decorated_function``, I presume things break as the inner function is typechecked?

I'm wondering since packages relying on ``@decorator_that_needs_to_be_bottom``need to be selective in their application of beartype anyway (if I am understanding correctly), could bottom be a better default?

**Comment by leycec** [comment]
**Gah!** So sorry for the extreme delay. April and May sure were wild rides, huh? :sweat_smile: 

Thankfully, @leycec is *finally* back in the bear saddle and takin' no QA prisoners. This is next up on the `TODO` list for @beartype 0.19.0. I must confess, though, that I don't quite grok your line of inquiry. That's probably more a "me" problem than a "you" problem. Autism: "It gets in the way again, huh?"

I will say that these configuration options are only a temporary salve. They're a cheap-and-easy convenience for users allergic to the current approach – but they don't particularly solve anything at all, really. It's trivial to envision real-world use cases in which the position of the `@beartype` decorator should contextually change depending on all of the *other* decorators present in a decorator chain. The `@beartype` decorator should reside at the top of a decorator chain in *some* cases, at the bottom in *other* cases, and somewhere in the middle in yet *other* cases.

Indeed, I ran some internal thought experiments that sufficiently horrified me. The worst case is truly far worse than any of us envisioned. Consider a hypothetical decorator `@transmute` that modifies the types hints of the callables it decorates. In this case, the only truly reliable means of type-checking a function decorated by both `@beartype` and `@transmute` is to **split the `@beartype` decorator into two distinct decorators:**

* `@beartype_params`, which type-checks *only* parameters.
* `@beartype_return`, which type-checks *only* returns.

`beartype.claw` import hooks (e.g., `beartype_this_package()`) would then need to sandwich decorators like `@transmute` between this pair of new @beartype decorators. Notably, this is what you'd need for @beartype to type-check the "original" parameters and returns as "originally" passed to that callable and returned from that callable:

```python
@beartype_params
@transmute
@beartype_return
def some_callable(param: SomeParamType) -> SomeReturnType: ...
```

Of course, that *really* sucks. That's a lot of work for both me and @beartype. Moreover, it's not particularly clear which third-party decorators are "transmuters." Clearly, `@jax.jit` is. But what about everything else? And how would those other decorators notify @beartype that @beartype needs to sandwich those decorators between a pair of decorators like above?

Ultimately, what's needed is "yet another PEP" standardizing some new miraculous form of **inter-decorator communication.** This PEP would enable *other* decorators to quietly notify the `@beartype` decorator of what the best position for `@beartype` should be in any given decorator chain.

But that PEP doesn't exist – and it probably never will. In the absence of that PEP, `@beartype` resorts to what `@beartype` does best: *makin' up everything as it goes along.*

**Comment by danielward27** [comment]
Sorry, I'll try to rephrase with a bit of code. So ``beartype`` at the top can sometimes avoid issues, e.g. with ``numba.jit``. But presumably, as soon as you start nesting functions, things would break? For example:
```python
from numba import jit
from beartype import beartype

@beartype
@jit
def my_jitted_inner(a: float, b: float):
    return a + b

@beartype
@jit
def my_jitted_outer(a: float, b: float):
    return my_jitted_inner(a, b)

my_jitted_outer(1., 2.)
```
So if you rely heavily on a decorator that needs to be below beartype, like ``numba.jit``, then things will inevitably break when you start nesting functions, meaning users would have to selectively apply beartype? If this was the case then I wondered if bottom should be the default, since 1) people are very bad at updating/forwarding types in decorators, 2) if the decorator has to go beneath beartype, then users likely couldn't use the import hook anyway, because any nesting of decorated functions would cause errors. Hope that makes more sense this time!

**Comment by leycec** [comment]
**Ah-ha!** I get you. Transitivity. Indeed, the order-dependency issues with decorator chaining are myriad.

That said... I'll probably avoid changing the default for the moment – but not because you're wrong. You're absolutely *not* wrong! But it's not clear that anyone's right, either. For [all the shoddy reasons delineated above](https://github.com/beartype/beartype/issues/374#issuecomment-2111500281), both of the existing options of "`@beartype` always above" and "`@beartype` always below" really blow chunks. :vomiting_face: 

There is no "one-size-fits-all" solution that even remotely addresses all possible edge cases here. Since both of the existing options fall down in obvious real-world scenarios, the "best" thing for @beartype to do (*...for the moment, anyway*) is just to **preserve backward compatibility.** If @beartype suddenly adopts "`@beartype` always below" as the new default, it's likely that we'd inadvertently break backward compatibility with everyone's code that currently works with "`@beartype` always above."

If someone's currently using @beartype, then (by definition) @beartype currently works for them – well enough, anyway. Preserving backward compatibility is *absolutely* the most important thing that @beartype can continue to deliver to everybody. That's *waaaaaaaaaay* more important than implementing new features. That's even more important than solving bugs.

That's why @beartype `0.18.0` and `0.18.1` were such catastrophic releases. They inadvertently broke backward compatibility for no particularly good reason and had to be permanently yanked from PyPI. It was the worst! I still have nightmares! I'm shuddering as we speak! And I'm petrified of repeating that disaster train.

But... *yeah.* I totally get you and @patrick-kidger. That's why I'm almost finished implementing this feature request for our upcoming @beartype 0.19.0 release. Ultimately, *you* and only *you* know what's best for your own codebase. Right? @beartype 0.19.0 respects your wisdom. Give everyone the power they so justly deserve, @beartype 0.19.0!

![](https://media1.tenor.com/m/6W_GlvUCMl8AAAAC/cat-lightning.gif)

**Comment by leycec** [comment]
**Resolved by ba3bb42116228019.** @beartype 0.19.0 unleashes your inner potential by allowing to replicate the @patrick-kidger Decorator Pattern, @danielward27. "But what *is* the @patrick-kidger Decorator Pattern!?", you may now be asking yourself.

Prepare to feel awe. This is the @patrick-kidger Decorator Pattern:

* When decorating classes, list runtime type-checking decorators like `@beartype` and `@jaxtyping.jaxtyped` *above* all other class decorators. This allows runtime type-checking decorators to handle standard decorators like `@dataclasses.dataclass`.
* When decorating callables, list runtime type-checking decorators like `@beartype` and `@jaxtyping.jaxtyped` *below* all other class decorators. Why? I'm waving my hands here.

In @beartype speak, this is the @patrick-kidger Decorator Pattern:

```python
from beartype import (
    BeartypeConf,
    BeartypeDecorationPosition,
)
from beartype.claw import beartype_this_package
beartype_this_package(
    conf=BeartypeConf(
        claw_decoration_position_funcs=BeartypeDecorationPosition.FIRST,
        claw_decoration_position_types=BeartypeDecorationPosition.LAST,
   ),
)
```

It is glorious! Nothing that has lived has ever been more beautiful.

You are now thinking this. Indeed, you are still thinking this should be the default. Unfortunately...

## I Was Dumb and Totally Forgot Why I Implemented This The Way I Did

Let's be honest: "Because backward compatibility" is the dumbest justification for anything ever. Sure, that's a *valid* justification – but it's not exactly convincing anyone of the correctness of its argument, is it?

Sadly, I just remembered a few hours ago why runtime type-checking decorators like `@beartype` and `@jaxtyping.jaxtyped` absolutely *should* be applied above all other decorators for both classes *and* callables. Why? Because PEP standards. If you apply a runtime type-checking decorator below other decorators, then you violate PEP standards – which is bad.

This includes the pivotal PEP 484-compliant `@typing.no_type_check` decorator. Consider this worst-case example:

```python
from beartype
from functools import cache
from typing import no_type_check

all_your_type_check_is_belong_to_us = no_type_check

@beartype
@all_your_type_check_is_belong_to_us
@cache
def this_day_all_type_checkers_die() -> int:
    return "This should be an integer, but it isn't. Too bad nobody even cares anymore, bro."

print(this_day_all_type_checkers_die())
```

The above example behaves as expected. The `@beartype` decorator is manually applied last and thus respects the preceding decorators – including the `@typing.no_type_check` decorator which has been maliciously renamed for no good reason. Users. What you gonna do, huh?

The above example does *not* behave as expected if you apply the @patrick-kidger Decorator Pattern. Doing so automatically applies the `@beartype` decorator last, which then ignores the preceding decorators, which then violates PEP standards – which (*...waitforit*) is bad.

## It Gets Worse, Because Of Course It Does

Users often manually decorate callables by the `@beartype` decorator configured with callable-specific beartype configurations. The usual example is instructing `@beartype` and *only* `@beartype` to ignore a given callable with the `BeartypeConf(strategy=BeartypeStrategy.O0)` configuration: e.g.,

```python
from beartype, BeartypeConf, BeartypeStrategy
from functools import cache

type_check_us_up_the_bomb = beartype(conf=BeartypeConf(strategy=BeartypeStrategy.O0))

@beartype
@type_check_us_up_the_bomb
@cache
def ill_get_you_and_your_little_type_checker_too() -> int:
    return "This should be an integer, but it isn't. I care. Why don't you? <-- totally not a leading question"

print(ill_get_you_and_your_little_type_checker_too())
```

If you apply the @patrick-kidger Decorator Pattern, then `@beartype` implicitly ignores *all* such configuration, which then causes everything across your entire codebase to break – which (*...waitforit*) is bad.

In short, I was dumb. I should have remembered all of this. But it's been *soooo* long since I first implemented any of this. Years. It's been years, guys. I am now old, bald, and wheezing in my sleep. My cats are old, bald, and wheezing in their sleep, too.

If I'd remembered any of this, I probably would have gently closed this feature request with the above explanation. Instead, I wasted everybody's time and implemented this. To be sure, **probably nobody should apply the @patrick-kidger Decorator Pattern.** It's cute, but it violates standards, ignores user configurations, and just generally falls down. A cute thing that falls down is no longer cute. *It's just sad.*

![](https://media1.tenor.com/m/wUxF2bvGoCwAAAAd/panda-fat.gif)
<sup>sad</sup>
